File: algorithms/07_typescript/math/brent_root_finding.ts
/**
 * Brent's Method for Root Finding (Numerical Recipes 3rd Ed. Chapter 9.3)
 * Guaranteed convergence combining Bisection, Secant, and Inverse Quadratic Interpolation.
 */

export function brentRoot(
  f: (x: number) => number,
  a: number,
  b: number,
  tol = 1e-12,
  maxIter = 100
): number {
  let fa = f(a);
  let fb = f(b);
  if (fa * fb > 0) throw new Error("Root is not bracketed in [a, b]");

  if (Math.abs(fa) < Math.abs(fb)) {
    [a, b] = [b, a];
    [fa, fb] = [fb, fa];
  }

  let c = a;
  let fc = fa;
  let mflag = true;
  let s = b;
  let fs = fb;
  let d = 0;

  for (let iter = 0; iter < maxIter; iter++) {
    if (Math.abs(fb) < tol || Math.abs(b - a) < tol) return b;

    if (fa !== fc && fb !== fc) {
      // Inverse quadratic interpolation
      s =
        (a * fb * fc) / ((fa - fb) * (fa - fc)) +
        (b * fa * fc) / ((fb - fa) * (fb - fc)) +
        (c * fa * fb) / ((fc - fa) * (fc - fb));
    } else {
      // Secant method
      s = b - fb * ((b - a) / (fb - fa));
    }

    const cond1 = (s - (3 * a + b) / 4) * (s - b) > 0;
    const cond2 = mflag && Math.abs(s - b) >= Math.abs(b - c) / 2;
    const cond3 = !mflag && Math.abs(s - b) >= Math.abs(c - d) / 2;
    const cond4 = mflag && Math.abs(b - c) < tol;
    const cond5 = !mflag && Math.abs(c - d) < tol;

    if (cond1 || cond2 || cond3 || cond4 || cond5) {
      s = (a + b) / 2;
      mflag = true;
    } else {
      mflag = false;
    }

    fs = f(s);
    d = c;
    c = b;
    fc = fb;

    if (fa * fs < 0) {
      b = s;
      fb = fs;
    } else {
      a = s;
      fa = fs;
    }

    if (Math.abs(fa) < Math.abs(fb)) {
      [a, b] = [b, a];
      [fa, fb] = [fb, fa];
    }
  }
  return b;
}

const root = brentRoot((x) => x ** 3 - x - 2, 1, 2);
if (Math.abs(root - 1.5213797068) > 1e-6) throw new Error("Brent root finding failed");
console.log(`NR Brent's Root Finding verified: root=${root}`);

File: algorithms/02_c/scientific_c/brent_root_finding.c
#include "brent_root_finding.h"
#include <math.h>

double brent_find_root(scalar_fn f, double a, double b, double tol, int max_iter) {
    double fa = f(a);
    double fb = f(b);
    if (fa * fb > 0) return a; // Root not bracketed
    
    double c = a, fc = fa;
    double d = b - a, e = d;
    
    for (int iter = 0; iter < max_iter; ++iter) {
        if (fb * fc > 0) {
            c = a; fc = fa;
            d = b - a; e = d;
        }
        if (fabs(fc) < fabs(fb)) {
            a = b; b = c; c = a;
            fa = fb; fb = fc; fc = fa;
        }
        double tol1 = 2.0 * 1e-15 * fabs(b) + 0.5 * tol;
        double xm = 0.5 * (c - b);
        if (fabs(xm) <= tol1 || fb == 0.0) return b;
        
        if (fabs(e) >= tol1 && fabs(fa) > fabs(fb)) {
            double s = fb / fa;
            double p, q;
            if (a == c) {
                p = 2.0 * xm * s;
                q = 1.0 - s;
            } else {
                q = fa / fc;
                double r = fb / fc;
                p = s * (2.0 * xm * q * (q - r) - (b - a) * (r - 1.0));
                q = (q - 1.0) * (r - 1.0) * (s - 1.0);
            }
            if (p > 0.0) q = -q;
            p = fabs(p);
            if (2.0 * p < fmin(3.0 * xm * q - fabs(tol1 * q), fabs(e * q))) {
                e = d;
                d = p / q;
            } else {
                d = xm; e = d;
            }
        } else {
            d = xm; e = d;
        }
        a = b; fa = fb;
        b += (fabs(d) > tol1) ? d : ((xm >= 0) ? tol1 : -tol1);
        fb = f(b);
    }
    return b;
}
